components/Buttons/src/MDCFloatingButton+Animation.m

```diff
@@ -0,0 +1,242 @@
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+#import "MDCFloatingButton+Animation.h"
+
+#if TARGET_IPHONE_SIMULATOR
+float UIAnimationDragCoefficient(void);  // Private API for simulator animation speed
+#endif
+
+static NSString *const kMDCFloatingButtonTransformKey = @"kMDCFloatingButtonTransformKey";
+static NSString *const kMDCFloatingButtonOpacityKey = @"kMDCFloatingButtonOpacityKey";
+
+static const CGFloat kMDCFloatingButtonTransformScale = 0.00001f;
+
+static const NSTimeInterval kMDCFloatingButtonEnterDuration = 0.270f;
+static const NSTimeInterval kMDCFloatingButtonExitDuration = 0.180f;
+
+static const NSTimeInterval kMDCFloatingButtonEnterIconDuration = 0.180f;
+static const NSTimeInterval kMDCFloatingButtonEnterIconOffset = 0.090f;
+static const NSTimeInterval kMDCFloatingButtonExitIconDuration = 0.135f;
+static const NSTimeInterval kMDCFloatingButtonExitIconOffset = 0.000f;
+
+static const NSTimeInterval kMDCFloatingButtonOpacityDuration = 0.015f;
+static const NSTimeInterval kMDCFloatingButtonOpacityEnterOffset = 0.030f;
+static const NSTimeInterval kMDCFloatingButtonOpacityExitOffset = 0.150f;
+
+@implementation MDCFloatingButton (Animation)
+
++ (CABasicAnimation *)animationWithKeypath:(nonnull NSString *)keyPath
+                                   toValue:(nonnull id)toValue
+                                 fromValue:(nullable id)fromValue
+                            timingFunction:(nonnull CAMediaTimingFunction *)timingFunction
+                                  fillMode:(nonnull NSString *)fillMode
+                                  duration:(NSTimeInterval)duration
+                               beginOffset:(NSTimeInterval)beginOffset {
+  CABasicAnimation *animation = [CABasicAnimation animationWithKeyPath:keyPath];
+  animation.toValue = toValue;
+  animation.fromValue = fromValue;
+  animation.timingFunction = timingFunction;
+  animation.fillMode = fillMode;
+  animation.removedOnCompletion = NO;
+  animation.duration = duration;
+  if (fabs(beginOffset) > DBL_EPSILON) {
+    animation.beginTime = CACurrentMediaTime() + beginOffset;
+  }
+
+#if TARGET_IPHONE_SIMULATOR
+  animation.duration *= [self fab_dragCoefficient];
+  if (fabs(beginOffset) > DBL_EPSILON) {
+    animation.beginTime = CACurrentMediaTime() + (beginOffset * [self fab_dragCoefficient]);
+  }
+#endif
+
+  return animation;
+}
+
+#if TARGET_IPHONE_SIMULATOR
++ (float)fab_dragCoefficient {
```

Comment:
The purpose is to slow-down our custom Core Animation animations to match those of UIViews.  In the simulator, turning on Slow Animation changes this value.  Since it's a private API we don't want it compiled into "real" targets.  I stole the code from the FlexibleHeaderView.